eclair-core/src/main/scala/fr/acinq/eclair/blockchain/electrum/ElectrumWallet.scala

```diff
@@ -971,35 +945,61 @@ object ElectrumWallet {
+        val tx1 = addUtxosWithDummySig(tx, candidates)
+        val tx2 = change.map(o => tx1.addOutput(o)).getOrElse(tx1)
         Transactions.weight2fee(feeRatePerKw, tx2.weight())
       }
 
-      // add change output only if non-dust, otherwise change is added to the fee
-      val (tx2, fee2, pos) = (spent - amount - fee1) match {
-        case dustChange if dustChange < dustLimit => (tx1, fee1 + dustChange, -1) // if change is below dust we add it to fees
-        case change => (tx1.addOutput(TxOut(change, computePublicKeyScript(currentChangeKey.publicKey))), fee1, 1) // change output index is always 1
+      val dummyChange = TxOut(Satoshi(0), computePublicKeyScript(currentChangeKey.publicKey))
+
+      @tailrec
+      def loop(current: Seq[Utxo], remaining: Seq[Utxo]) : (Seq[Utxo], Option[TxOut]) = {
+        totalAmount(current) match {
+          case total if total - amount - computeFee(current, None) < Satoshi(0) && remaining.isEmpty =>
+            // not enough funds to send amount and pay fees even without a change output
+            throw new IllegalArgumentException("insufficient funds")
+          case total if total - amount - computeFee(current, None) < Satoshi(0) =>
+            // not enough funds, try with an additional input
+            loop(remaining.head +: current, remaining.tail)
+          case total if total - amount - computeFee(current, None) <= dustLimit =>
+            // change output would be below dust, we don't add one and just overpay fees
+            (current, None)
+          case total if total - amount - computeFee(current, Some(dummyChange)) <= dustLimit && remaining.isEmpty =>
+            // change output is above dust limit but cannot pay for it's own fee, and we have no more utxos => we overpay a bit
+            (current, None)
+          case total if total - amount - computeFee(current, Some(dummyChange)) <= dustLimit =>
+            // try with an additional input
+            loop(remaining.head +: current, remaining.tail)
+          case total =>
+            val fee = computeFee(current, Some(dummyChange))
+            val change = dummyChange.copy(amount = total - amount -fee)
+            (current, Some(change))
+        }
       }
 
+      val (selected, change_opt) = loop(Seq.empty[Utxo], unlocked2)
+
       // sign our tx
+      val tx1 = addUtxosWithDummySig(tx, selected)
+      val tx2 = change_opt.map(out => tx1.addOutput(out)).getOrElse(tx1)
       val tx3 = signTransaction(tx2)
-      //Transaction.correctlySpends(tx3, utxos.map(utxo => utxo.outPoint -> TxOut(Satoshi(utxo.item.value), computePublicKeyScript(utxo.key.publicKey))).toMap, ScriptFlags.STANDARD_SCRIPT_VERIFY_FLAGS)
 
-      // and add the completed tx to the lokcs
+      // and add the completed tx to the locks
       val data1 = this.copy(locks = this.locks + tx3)
-      val fee3 = spent - tx3.txOut.map(_.amount).sum
+      val fee = selected.map(s => Satoshi(s.item.value)).sum - tx3.txOut.map(_.amount).sum
```

Comment:
Why don't we put a `lazy val fee` inside `Transaction` that does that?
It seems like it's unnecessary to separately return the fee, it's a property of the `Transaction`.

Reply:
Because a transaction cannot compute its onchain fees since it is missing the values of the UTXOs it is spending, it just know their outpoints.

Reply:
Indeed, the outpoint doesn't have enough information to compute this, my bad.